### PR TITLE
 Add support for executing test classes in parallel via JUnit Platform

### PR DESCRIPTION
**Pull request by marcphilipp**
Prior to this PR, our listeners only supported a single executing test class at a time. That assumption has now been removed when running tests on the JUnit Platform. The listener is now completely independent from the one we use for JUnit 4 which should make it easier to work on other JUnit Platform issues in the future.

### CODE REVIEW

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junitplatform/JUnitPlatformIntegrationTest.groovy [hunk 1]
```diff
@@ -362,4 +365,49 @@ public class StaticInnerTest {
         'includeEngines' | '"junit-jupiter"'
         'excludeEngines' | '"junit-jupiter"'
     }
+
+    @Issue('https://github.com/gradle/gradle/issues/6453')
+    def "can handle parallel test execution"() {
+        given:
+        def numTestClasses = 2
```
Comment (oehme) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junitplatform/JUnitPlatformIntegrationTest.groovy]: Do 2 tests consistently reproduce the previous issue? I think a few dozen wouldn't make this test much longer and might trigger concurrency issues more reliably.

File: subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitIntegrationTest/junit3Tests/src/test/java/org/gradle/Junit3Test.java [hunk 2]
```diff
@@ -5,5 +5,6 @@
 public class Junit3Test extends TestCase {
     public void testRenamesItself() {
         setName("a test that renames itself");
+        fail("epic");
```
Comment (oehme) [inline subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitIntegrationTest/junit3Tests/src/test/java/org/gradle/Junit3Test.java]: 😆 

Why the change though?

File: subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java [hunk 3]
```diff
@@ -16,36 +16,47 @@
...
+import static org.gradle.api.internal.tasks.testing.junitplatform.VintageTestNameAdapter.isVintageDynamicLeafTest;
+import static org.gradle.api.internal.tasks.testing.junitplatform.VintageTestNameAdapter.vintageDynamicClassName;
+import static org.gradle.api.internal.tasks.testing.junitplatform.VintageTestNameAdapter.vintageDynamicMethodName;
+import static org.gradle.api.tasks.testing.TestResult.ResultType.SKIPPED;
+import static org.junit.platform.engine.TestDescriptor.Type.CONTAINER;
+import static org.junit.platform.engine.TestExecutionResult.Status.ABORTED;
+import static org.junit.platform.engine.TestExecutionResult.Status.FAILED;
 
 public class JUnitPlatformTestExecutionListener implements TestExecutionListener {
-    private final GenericJUnitTestEventAdapter<String> adapter;
```
Comment (oehme) [inline subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java]: Do you plan to remove this one and make JUnit 4 handling simpler as a result?

File: subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java [hunk 4]
```diff
@@ -60,119 +71,131 @@ public void testPlanExecutionFinished(TestPlan testPlan) {
...
-        if (isFailedContainer(testIdentifier, testExecutionResult)) {
-            // only leaf methods triggered start events previously
-            // so here we need to add the missing start events
-            adapter.testStarted(testIdentifier.getUniqueId(), getDescriptor(testIdentifier));
-            testFinished(testIdentifier, testExecutionResult);
-        } else if (isLeafTest(testIdentifier)) {
-            testFinished(testIdentifier, testExecutionResult);
+    private void reportStartedUnlessAlreadyStarted(TestIdentifier testIdentifier) {
+        // guard against edge cases (e.g. JUnit 4 classes with custom runners that report the class as ignored after reporting it as started)
+        if (!wasStarted(testIdentifier)) {
```
Comment (oehme) [inline subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java]: ❌ Isn't there a race condition here, where two threads can check "wasStarted" and then both decide to call "started()"?

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestListenerBuildOperationAdapterIntegrationTest.groovy [hunk 5]
```diff
@@ -56,17 +58,31 @@ class TestListenerBuildOperationAdapterIntegrationTest extends JUnitMultiVersion
...
-        ops[4].details.testDescriptor.composite == false
-
-        ops[5].details.testDescriptor.name == "ok"
-        ops[5].details.testDescriptor.className == "org.gradle.OtherTest"
-        ops[5].details.testDescriptor.composite == false
+        if (isVintage()) {
+            ops[3].details.testDescriptor.name == "org.gradle.OkTest"
+            ops[3].details.testDescriptor.className == "org.gradle.OkTest"
+            ops[3].details.testDescriptor.composite == true
+            offset++
```
Comment (oehme) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestListenerBuildOperationAdapterIntegrationTest.groovy]: Using an iterator may improve readability over index computations

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junitplatform/JUnitPlatformIntegrationTest.groovy (hunk 1 above)
Comment (marcphilipp) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junitplatform/JUnitPlatformIntegrationTest.groovy]: They did. But you're right, adding more doesn't hurt, so I will.

File: subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitIntegrationTest/junit3Tests/src/test/java/org/gradle/Junit3Test.java (hunk 2 above)
Comment (marcphilipp) [reply subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitIntegrationTest/junit3Tests/src/test/java/org/gradle/Junit3Test.java]: This is really an edge case and I wanted to make sure that we at least consistently report a failed test in both JUnit 4 and Vintage.

File: subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java (hunk 3 above)
Comment (marcphilipp) [reply subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java]: Good catch, will do!

Comment (marcphilipp) [reply subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java]: Done in 1a61bb4608aa62b85361a07b3e2d934bd32d9504.
